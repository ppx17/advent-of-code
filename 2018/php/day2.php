<?php
$data = $data ?? file_get_contents("../input/input-".basename(__FILE__, '.php').".txt");

$ids = explode("\n", trim($data));

function part1(array $ids): int
{
    $twos = $threes = 0;
    foreach ($ids as $id) {
        $counts = [];
        for ($i = 0; $i < strlen($id); $i++) {
            $counts[$id[$i]]++;
        }
        if (in_array(2, $counts)) {
            $twos++;
        }
        if (in_array(3, $counts)) {
            $threes++;
        }
    }

    return $twos * $threes;
}

function common($first, $second): string
{
    $result = '';
    for ($i = 0; $i < strlen($first); $i++) {
        $result .= ($first[$i] === $second[$i]) ? $first[$i] : '';
    }
    return $result;
}

function part2($ids): ?string
{
    for ($fi = 0; $fi < count($ids); $fi++) {
        for ($si = $fi + 1; $si < count($ids); $si++) {
            if (levenshtein($ids[$fi], $ids[$si]) === 1) {
                return common($ids[$fi], $ids[$si]);
            }
        }
    }
    return null;
}
echo "part1: " . part1($ids) . PHP_EOL;
echo "part2: " . part2($ids);