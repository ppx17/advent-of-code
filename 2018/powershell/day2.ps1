Param(
    [string]$InputFile = '../input/input-day2.txt'
);

$Ids = (Get-Content $InputFile);

$Twos = $Threes = 0;

foreach($Id in $Ids) {
    $Counts = $Id.ToCharArray() | Group-Object | Select-Object -ExpandProperty Count;
    if($Counts -contains 2) {
        $Twos++;
    }
    if($Counts -contains 3) {
        $Threes++;
    }
}

$Part1 = $Twos * $Threes;

Write-Output "Part 1: ${Part1}";

Class Part2Solver {
    [int]difference([string]$First, [string]$Second) {
        $diffCount = 0;
        for($i=0; $i -lt $First.Length; $i++) {
            if($First[$i] -ne $Second[$i]) {
                $diffCount++;
            }
        }
        return $diffCount;
    }

    [string]common([string]$First, [string]$Second) {
        $Result = "";
        for($i=0; $i -lt $First.Length; $i++) {
            if($First[$i] -eq $Second[$i]) {
                $Result += $First[$i];
            }
        }
        return $Result;
    }

    [string]Solve([string[]] $Ids) {
        for($fi = 0; $fi -lt $Ids.Length; $fi++) {
            for($si = $fi + 1; $si -lt $Ids.Length; $si++) {
                $Difference = $this.difference($Ids[$fi], $Ids[$si]);
                if($Difference -eq 1) {
                    return $this.common($Ids[$fi], $Ids[$si]);
                }
            }
        }
        return $null;
    }
}

$Solver = [Part2Solver]::new();
$Part2 = $Solver.Solve($Ids);
Write-Output ("Part2: ${Part2}");
