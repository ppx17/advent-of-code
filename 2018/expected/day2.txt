Part 1: 7904
Part 2: wugbihckpoymcpaxefotvdzns
